### PR TITLE
[WIP] Python3 Port

### PR DESCRIPTION
**Pull request by SteadBytes**
Part of Minecraft-Overviewer Python3 port https://github.com/overviewer/Minecraft-Overviewer/pull/1540

Aim to upgrade Buildbot + all builders to Python3 and improve local development setup

- [x] Update scripts/Dockerfiles for newer package versions (gpg, rpmsign etc)
- [x] Update Buildbot to latest + Python3
- [ ] Update builders:
  - [ ] centos6-64 (awaiting EPEL 6 Python3 Pillow package)
  - [ ] centos6-32 (awaiting EPEL 6 Python3 Pillow package)
  - [x] centos7-64
  -  :no_entry_sign:  centos7-32 removed due to no EPEL 7 32 bit repo
  - [x] stretch-64
  - [x] stretch-32
  - [x] EC2 windows AMI
    - [x] Updated `ShellCommand`s to use Python3 exe
    - Don't have access to the actual AMI's used, will require core dev to update
- [x] Improve local development:
  - [x] Mount config/scripts as volumes to enable changes without rebuilding images
  - [x] Toggle EC2 builders
  - [x] Ensure Buildbot config references local resources (i.e. local Flask app instead of overviewer.org)
  - [x] Ensure Buildbot auth can be used locally with a non-core dev Github account
- [x] Documentation

## Authentication Changes

Buildbot authentication has changed significantly in newer releases, as such the `OverviewerAuth` class ([source](https://github.com/overviewer/Overviewer-Web/blob/master/buildbot/master.cfg#L71)) which uses the Flask session user as authentication no longer works and I was unable to find a good way to implement it. It has been replaced with the built in `GitHubAuth` and is configured to allow only members of the `overviewer` organisation to run builds and make changes (same behaviour as before).

The only downside is that a user will need to re-authenticate when accessing the build system, even if they have previously authenticated with https://overviewer.org

### CODE REVIEW

**Comment by SteadBytes** [comment]
@CounterPillow , in order to upgrade Buildbot (v0.8.x) to the current version (v2.x) requires starting with a clean database. From the [docs](http://docs.buildbot.net/current/manual/upgrading/nine-upgrade.html):

> First important note is that Buildbot does not support an upgrade of a 0.8.x instance to 0.9.x. Notably the build data and logs will not be accessible anymore if you upgraded, thus the database migration scripts have been dropped.

Attempting to upgrade the database with `buildbot upgrade-master` produces the following error message:
```
buildbot.db.model.EightUpgradeError: You are trying to upgrade a buildbot 0.8.x master to buildbot 0.9.x
This is not supported. Please start from a clean database
http://docs.buildbot.net/latest/manual/installation/nine-upgrade.html
```

All the desired behaviour of the build system can be kept (and likely improved), however we will have to drop the historical data I think. Is this ok?

**Comment by CounterPillow** [comment]
@SteadBytes Seems ok to me, AFAIK the builds are kept outside of the database anyway so the historical value of the db is very little as it's mostly just automatic build triggered messages.

**Comment by SteadBytes** [comment]
Great, thanks @CounterPillow . I will continue upgrading to the latest version :+1:

**Comment by SteadBytes** [comment]
@CounterPillow I've hit a problem with the RPM Build and Install requirements. `python-imaging` and `python-imaging-devel` are required [here](https://github.com/overviewer/Overviewer-Web/blob/ca12755ba582e0df0bac7d2726ed1197b368c05d/buildbot/rpm/control/centos7/Minecraft-Overviewer.spec#L11), however there is no `python3-imaging` packages for CentOS (that I can find) like there is on Debian.

When building from source this can be overcome by installing those from source prior to building overviewer, however that won't work for an RPM package as it can't be specified as a dependency. Do you have any ideas on how to get around this? :thinking:

**Comment by CounterPillow** [comment]
@SteadBytes I was under the assumption that those would be packaged by [EPEL](https://fedoraproject.org/wiki/EPEL), but apparently they're not. I'm not really sure how we could proceed. Maybe we should just drop the RPM repos to avoid them holding us back?

**Comment by CounterPillow** [comment]
There are two other alternatives:

1. Get a Python 3 pillow package into EPEL for at least CentOS 7
2. Provide a Python 3 pillow package in our own repository which we maintain

I'm not really sure I like the second option since pillow is a reasonably widely depended upon package for people to not want some random Minecraft project to provide them with one for their entire system.

**Comment by SteadBytes** [comment]
@CounterPillow , I agree that the first of those options is preferable. Is it worth waiting until we can get that into EPEL or to continue without CentOS builds for now until that happens?

**Comment by CounterPillow** [comment]
@SteadBytes I've [filed a request with EPEL](https://bugzilla.redhat.com/show_bug.cgi?id=1701767), we'll see how it goes over the next few days.

**Comment by SteadBytes** [comment]
@CounterPillow Ok great thank you, I'll carry on with the rest for now and I can pick up the CentOS builds after that 👍

**Comment by SteadBytes** [comment]
@CounterPillow , this is mostly complete now - apart from the CentOS builds of course. I have left the CentOS dockers and RPM files in for now as hopefully we will be able to include them.

I don't have access to the AWS Windows AMI's used so am unable to update those as needed however the `ShellCommand`'s in the Builbot config will now use the Python3 exe (though unable to test this change). Please let me know if you have any suggestions/comments :smile:

**Comment by agrif** [comment]
The windows AMIs are my fault. I can test this and make any changes needed before we merge this. The changes look fine, though.

**Comment by SteadBytes** [comment]
Thanks a lot @agrif, let me know if there's anything that needs changing 👍

**Comment by SteadBytes** [comment]
@CounterPillow Doesn't look like there's any progress on the EPEL request for python3 Pillow :thinking: Any ideas on how we should proceed?

**Comment by CounterPillow** [comment]
One of the dependencies for Pillow will be pushed to the EPEL stable repos in 5 days time [according to their updates system](https://bodhi.fedoraproject.org/updates/python-sphinx_rtd_theme-0.2.4-3.el7), whether or not this means they'll immediately start working on a python3 compatible Pillow build remains to be seen.

In the meantime we can proceed without CentOS builds; if CentOS users want to run outdated software then I guess that's what they get. I've talked to agrif about the Windows AMI stuff and he told me I should poke him thursday or on the weekend and he'll work on it (and presumably also on merging this).

**Comment by SteadBytes** [comment]
Ok that sounds good - hopefully they will start working on python3 Pillow build soon :crossed_fingers: Is it worth removing the CentOS docker containers from the `docker-compose.yml` for now as there's no point in starting them when BuildBot won't be utilising them?

Great, thank you for poking @agrif :grin: haha and for reviewing/assisting in this PR.

**Comment by agrif** [comment]
We should keep them around, in hope of a brighter future!

(Also they're not really hurting anything.)

**Comment by SteadBytes** [comment]
Good idea @agrif !

File: buildbot/master/master.cfg [hunk 1]
```diff
@@ -462,26 +469,31 @@ def rpm(rpmbase, mockbase, mockconfig, rpmarch, mockarch):
 
     # link the new rpm into the repo
     system_map = {'centos6': '6', 'centos7': '7'}
-    rpm_package_area = RPM_REPO + "/" + system_map[rpmbase] + "/" + rpmarch + "/packages"
+    rpm_package_area = RPM_REPO / system_map[rpmbase] / + rpmarch / "packages"
```
Comment (agrif) [inline buildbot/master/master.cfg]: I know the RPM builders are untested, but wouldn't this:

    + rpmarch

be a TypeError? I don't think unary `+` works on strings or paths.

**Review by agrif — COMMENTED** [review]
Gave this a once over and got a windows AMI ready. Might be merging this this weekend, depending. Thanks again for all the hard work!

File: buildbot/master/master.cfg (hunk 1 above)
Comment (SteadBytes) [reply buildbot/master/master.cfg]: Ah yes thank you for spotting that! It was just meant to be `/ rpmarch`, silly typo I'll fix it now :+1:

**Comment by SteadBytes** [comment]
@agrif That's great thank you very much! You are more than welcome :smile:

File: buildbot/master/master.cfg (hunk 1 above)
Comment (SteadBytes) [reply buildbot/master/master.cfg]: Done :smile:

**Comment by CounterPillow** [comment]
An EPEL7 python3-pillow is now in testing https://bugzilla.redhat.com/show_bug.cgi?id=1701767#c4

**Comment by SteadBytes** [comment]
Do you think it's worth trying it out from the testing repo? I can update the CentOS builds to use it so we can be sure the build process works and then switch to fetching it from the normal repo once it's out of testing?

**Comment by CounterPillow** [comment]
If you want to try making the builders use it, I could add positive karma to the testing package if Overviewer builds and runs with it. Though enabling epel-testing in this PR doesn't seem too wise if it's just a matter of ~10 days until this is in stable.

Once it's available in stable, the Great Mergening can begin, so I'm positive we'll be on Python 3 in the next 3 weeks.

**Comment by SteadBytes** [comment]
@CounterPillow  ok great, if I get a chance I'll try and get it working just to see. Otherwise, I'll get this PR updated to be ready for when it's in stable (updating the CentOS dockerfiles and RPM configs) :+1: Exciting times ahead!

**Comment by SteadBytes** [comment]
Looks like we'll also need to get `python3-numpy` into EPEL too, there's an existing testing package https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2019-543c65499f but it's got no Karma. However, there does currently exist a `python36-numpy` which works for building against Python 3.6 - I'll go with using this for now.

We do however need `python3-pillow-devel` as the `Imaging.h` header files are required for building from source :thinking:

**Comment by CounterPillow** [comment]
Admittedly we could just download Imaging.h ourselves and point the PIL_INCLUDE_DIR at it, since it's only needed for compiling and not running.

**Comment by SteadBytes** [comment]
Good point, I guess that would have to be done in the `%setup` section of
the RPM specs?

**Comment by CounterPillow** [comment]
`%setup` appears to be a macro for downloading and unpacking source archives, the section this is done in is `%prep` apparently. So if we want to go down the route of doing it in the RPM spec, a wget or curl in `%prep` seems like the proper way to go. On the other hand, we could also just wget/curl it in the dockerfile or something.

**Comment by CounterPillow** [comment]
Okay, the python3-pillow package is now in stable EPEL7.

**Comment by SteadBytes** [comment]
I think doing it in `%prep` in the RPM spec is the way to go. Just adding it in the dockerfile would mean that other people would be unable to build the RPM solely using the spec. I've already got a branch off of this with `python-3-pillow` installation/usage in preparation for when it got to stable. I'll add in the `%prep` step and merge it into this and then we should be good to go :smile:

**Comment by SteadBytes** [comment]
I don't think we are able to build for CentOS 7 32 bit, there are no EPEL-7 32 bit repos available https://mirrors.fedoraproject.org/mirrorlist?repo=epel-7&arch=i386 . 32 bit for CentOS 6 is fine of course. Is this a major issue? I can't really see a way around it if it is :thinking:

**Comment by SteadBytes** [comment]
Doh! Of course, to build either of the CentOS 6 packages we'll need `python3-pillow` in EPEL 6 too :sweat: At the moment, the only CentOS version we can build is CentOS 7 x86. Is it worth creating another issue similar to https://bugzilla.redhat.com/show_bug.cgi?id=1701767 but for EPEL 6?

**Comment by SteadBytes** [comment]
CentOS 7 64 bit is now working using EPEL 7 `python3-pillow` package, cloning Pillow repo into `%{_tmppath}` directory in %prep and setting `PIL_INCLUDE_DIR` to `%{_tmppath}/src/libImaging` during `%build`. I have removed the CentOS 7 32 bit builder due to the previously mentioned missing EPEL 7 32 bit repos. The CentOS 6 builders are set up but do still require `python3-pillow` to be in the EPEL 6 repos

**Comment by SteadBytes** [comment]
I have filed [another request](https://bugzilla.redhat.com/show_bug.cgi?id=1718518) for an EPEL 6 `python3-pillow` package :crossed_fingers:

**Comment by CounterPillow** [comment]
I'm fine with dropping CentOS 6 completely and going ahead right now.

@agrif what's the status on the Windows AMI stuff?

I also guess we should first merge this, then merge the Python 3 PR in the overviewer repo, and then make a blog post telling people about it.

**Comment by agrif** [comment]
@CounterPillow windows AMIs are basically good to go, though I do remember I have a short to-do list there after merge, so I should probably be the one to merge it.

That can be this week, if you guys are ready. Is the Python 3 port of Overviewer also done?

**Comment by CounterPillow** [comment]
Yes, the Python 3 port of Overviewer is done, pending a merge.

**Comment by SteadBytes** [comment]
@CounterPillow Ok, if `python3-pillow` gets into EPEL 6 then I'll re-introduce the CentOS 6 builds . For now, I'll disable them (or remove them entirely if preferred?)

**Comment by agrif** [comment]
Ok, current status:

 * It starts the EC2 builders but they never connect. I have no idea why.
 * Signing anything with gpg fails with `gpg: signing failed: Inappropriate ioctl for device` or `gpg: signing failed: End of file` and I don't know why.

If those pain points can be resolved, I can merge this. Maybe I'll figure them out tomorrow.

**Comment by SteadBytes** [comment]
I'm not sure about the EC2 builders issue as I don't know the deployment set up. However I'll throw out some ideas: are any of the EC2 instances in a VPC? Do the EC2 instances have the correct security groups for inbound/outbound traffic? Are there any firewall rules blocking the traffic (either on Buuldbot server or EC2 builders).

The `gpg` issue I have seen before when first working on this PR. Its caused by `gpg` not being able to read input  from the expected source (`stdin`). I noticed you enabled batch mode and have a feeling that could be the cause, could you please clarify why that was added? However I'll do some further digging if I get a chance today 😊

**Comment by SteadBytes** [comment]
@agrif Does signing fail for any specific builders? I can't reproduce it locally :thinking:

**Comment by agrif** [comment]
@SteadBytes I see the `ioctl` error on both debian and the centos builders. I see the `End of file` error in bbmaster, when it boots up and runs make in the debian repo.

I added `--batch` to `gpg-agent` under the suspicion that it was trying to run pinentry and failing because it didn't have a tty. It didn't end up helping, though. I got the same errors in the builders before and after.

**Comment by agrif** [comment]
Ok, some more details: running gpg will attempt to ask for a passphrase, even after starting an agent and running `gpg-preset-passphrase`. If I run `gpg-agent` without `--batch`, it will fail with the `ioctl` error every time as it tries to open pinentry on a tty it doesn't have.

If I run `gpg-agent` with `--batch`, it will know it can't open a pinentry on the first call to `gpg`, so instead it crashes (???) and gives the `End of file` error. Subsequent calls to `gpg` run without an agent, attempt to launch pinentry, and fail again with the `ioctl` error.

I've confirmed in an interactive tty that it's the correct passphrase for the key. I honestly have no idea what is going on.

**Comment by agrif** [comment]
Aha! My fault. I was setting the `KEYGRIP` env var to the fingerprint, not the key grip. I guess it's odd that it ever worked before this branch, but whatever. It works now!

**Comment by agrif** [comment]
Ok, that cleans up the big things. I think we'll need Ubuntu builders (since they run a different version of python) but for now it works.

Thank you so much for your hard work!

**Comment by SteadBytes** [comment]
Woohoo! Glad you got it sorted out 😁 Python3 is finally here! You're more than welcome, thank you for the opportunity!

**Pull request merged**